internal/gatewayapi/route.go

```diff
@@ -1160,3 +1189,98 @@ func getIREndpointsFromEndpointSlice(endpointSlices []*discoveryv1.EndpointSlice
 
 	return endpoints
 }
+
+func GetTargetBackendReference(backendRef gwapiv1a1.BackendRef, namespace string) gwapiv1a1.PolicyTargetReferenceWithSectionName {
+	ref := gwapiv1a1.PolicyTargetReferenceWithSectionName{
+		PolicyTargetReference: gwapiv1a1.PolicyTargetReference{
+			Group:     *backendRef.Group,
+			Kind:      *backendRef.Kind,
+			Name:      backendRef.Name,
+			Namespace: NamespacePtr(NamespaceDerefOr(backendRef.Namespace, namespace)),
+		},
+		SectionName: SectionNamePtr(strconv.Itoa(int(*backendRef.Port))),
+	}
+	return ref
+}
+
+func TargetMatched(policy gwapiv1a1.BackendTLSPolicy, target gwapiv1a1.PolicyTargetReferenceWithSectionName) bool {
+
+	policyTarget := policy.Spec.TargetRef
+
+	if target.Group == policyTarget.Group &&
+		target.Kind == policyTarget.Kind &&
+		target.Name == policyTarget.Name &&
+		NamespaceDerefOr(policyTarget.Namespace, policy.Namespace) == string(*target.Namespace) {
+		if policyTarget.SectionName != nil && *policyTarget.SectionName != *target.SectionName {
+			return false
+		}
+		return true
+	}
+	return false
+}
+
+func getBackendTLSPolicy(policies []*gwapiv1a1.BackendTLSPolicy, target gwapiv1a1.PolicyTargetReferenceWithSectionName) *gwapiv1a1.BackendTLSPolicy {
+	for _, policy := range policies {
+		if TargetMatched(*policy, target) {
+			return policy
+		}
+	}
+	return nil
+}
+
+func getBackendTLSBundle(policies []*gwapiv1a1.BackendTLSPolicy, configmaps []*corev1.ConfigMap, secrets []*corev1.Secret, targetBackend gwapiv1a1.PolicyTargetReferenceWithSectionName) (error, *ir.TLSBundle) {
+
+	backendTLSPolicy := getBackendTLSPolicy(policies, targetBackend)
+
+	if backendTLSPolicy == nil {
+		return nil, nil
+	}
+
+	tlsBundle := &ir.TLSBundle{}
+
+	caRefMap := make(map[string]string)
+
+	for _, caRef := range backendTLSPolicy.Spec.TLS.CACertRefs {
+		caRefMap[string(caRef.Name)] = string(caRef.Kind)
+	}
+
+	ca := ""
+
+	for _, cmap := range configmaps {
+		if kind, ok := caRefMap[cmap.Name]; ok && kind == cmap.Kind {
+			if crt, dataOk := cmap.Data["ca.crt"]; dataOk {
+				if ca != "" {
+					ca += "\n"
+				}
+				ca += crt
+			} else {
+				return fmt.Errorf("no ca found in configmap %s", cmap.Name), nil
+			}
+		}
+	}
+
+	for _, secret := range secrets {
+		if kind, ok := caRefMap[secret.Name]; ok && kind == secret.Kind {
+			if tls, tlsOk := secret.Data["tls.crt"]; tlsOk {
+				if key, keyOk := secret.Data["tls.key"]; keyOk {
+					tlsBundle.CertificateByte = tls
+					tlsBundle.PrivateKeyByte = key
+				} else {
+					return fmt.Errorf("no tls key found in secret %s", secret.Name), nil
+				}
+			}
+		}
+	}
+
+	if ca == "" {
+		return fmt.Errorf("no ca found in referred configmaps"), nil
```

Comment:
I saw the doc of gateway api said as below:

	// Support: Core - An optional single reference to a Kubernetes ConfigMap,
	// with the CA certificate in a key named `ca.crt`.
https://github.com/kubernetes-sigs/gateway-api/blob/54387b1e8db24e0c3eb644fa6c48a1b9ee711b39/apis/v1alpha2/backendtlspolicy_types.go#L97-L98

Does that mean the CA can be optional? If the CA isn't here, it means we needn't validate the backend service's cert.

Reply:
@soulxu if you are using upstreamTLS you must provide ca as it is the core support. if the ca is ok, then we can look for the cert and key for mtls and which is optional .